Add PodMonitor

auhlig opened the pull request:
As described in #38.
Still work in progess. To be continued after easter holidays

auhlig commented:
Could you take a look @brancz, @metalmatze?

brancz commented on Documentation/api.md:
```diff
@@ -202,6 +205,44 @@ NamespaceSelector is a selector for selecting either all namespaces or a list of
+| Field | Description | Scheme | Required |
+| ----- | ----------- | ------ | -------- |
+| jobLabel | The label to use to retrieve the job name from. | string | false |
+| targetLabels | TargetLabels transfers labels on the Kubernetes Service onto the target. | []string | false |
```
this should just be what `podTargetLabels` is no?

auhlig replied:
Yep. Will change

brancz commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -534,16 +575,17 @@ func TestPodTargetLabels(t *testing.T) {
-		map[string]*monitoringv1.ServiceMonitor{
-			"testservicemonitor1": &monitoringv1.ServiceMonitor{
+		nil,
+		map[string]*monitoringv1.PodMonitor{
```
We still want the test that was here previously. Could you extract this into a separate test?

auhlig replied:
Sure

brancz commented on Documentation/api.md:
```diff
@@ -202,6 +205,44 @@ NamespaceSelector is a selector for selecting either all namespaces or a list of
+| jobLabel | The label to use to retrieve the job name from. | string | false |
+| targetLabels | TargetLabels transfers labels on the Kubernetes Service onto the target. | []string | false |
+| podTargetLabels | PodTargetLabels transfers labels on the Kubernetes Pod onto the target. | []string | false |
+| endpoints | A list of endpoints allowed as part of this PodMonitor. | [][Endpoint](#endpoint) | true |
```
This is a little difficult, at this point I'm not comfortable introducing everything the `Endpoint` struct offers in terms of configuration, for the same reasons why https://github.com/coreos/prometheus-operator/pull/2524 is happening. I'd prefer to have a duplicate struct for this, that does not have the problematic fields that are the goal to be deprecated as part of the overall effort of https://github.com/coreos/prometheus-operator/pull/2524. I'd suggest that we just copy the current `Endpoint` struct and for now remove the `TLSConfig` and basic auth options.

squat replied:
Yes agreed. We may also want to be cautious here in future naming as `endpoints` is an existing term in the K8s world and is closely related to the concept of services.

brancz replied:
Maybe `PodMetricsEndpoint` ?

auhlig replied:
`PodMetricsEndpoint` sounds good

squat commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -542,6 +551,38 @@ type ServiceMonitorSpec struct {
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#metadata
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of desired Pod selection for target discrovery by Prometheus.
```
s/discrovery/discovery/

auhlig replied:
done

squat commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -542,6 +551,38 @@ type ServiceMonitorSpec struct {
+type PodMonitorSpec struct {
+	// The label to use to retrieve the job name from.
+	JobLabel string `json:"jobLabel,omitempty"`
+	// TargetLabels transfers labels on the Kubernetes Service onto the target.
```
This should be agnostic of a Kubernetes Service. Do we mean `Kubernetes Pod` here instead?

auhlig replied:
yep. fixed.

squat commented on pkg/prometheus/operator.go:
```diff
@@ -1437,6 +1499,51 @@ func (c *Operator) selectServiceMonitors(p *monitoringv1.Prometheus) (map[string
+	}
+
+	podMonitors := []string{}
+	for k := range res {
```
Instead of looping again here, we can add the keys to the slice on line 1533, when we are already looping.

auhlig replied:
✅

squat commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,274 @@ func (cg *configGenerator) generateConfig(
+	if version.Major == 1 && version.Minor < 7 {
+		// Filter targets based on the namespace selection configuration.
+		// By default we only discover services within the namespace of the
+		// ServiceMonitor.
```
s/ServiceMonitor/PodMonitor/g

auhlig replied:
✅

squat commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,274 @@ func (cg *configGenerator) generateConfig(
+
+	// By default, generate a safe job name from the pod name.  We also keep
+	// this around if a jobLabel is set in case the targets don't actually have a
+	// value for it. A single service may potentially have multiple metrics
```
s/service/pod/

auhlig replied:
✅

auhlig commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -583,6 +581,41 @@ type PodMonitorSpec struct {
 
+// PodMetricsEndpoint defines a scrapeable endpoint of a Kubernetes Pod serving Prometheus metrics.
+// +k8s:openapi-gen=true
+type PodMetricsEndpoint struct {
```
The `PodMetricsEndpoint` for a `PodMonitor` copies the `Endpoint` struct used in a `ServiceMonitor`. This is used to dis-entangle both. Ok with this @brancz?

brancz commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -542,6 +551,71 @@ type ServiceMonitorSpec struct {
+	// Timeout after which the scrape is ended
+	ScrapeTimeout string `json:"scrapeTimeout,omitempty"`
+	// TLS configuration to use when scraping the endpoint
+	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
```
I'd prefer to leave this out for now, due to the previously mentioned reasons.

auhlig replied:
right. removed

brancz commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -542,6 +551,71 @@ type ServiceMonitorSpec struct {
+	// TLS configuration to use when scraping the endpoint
+	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
+	// File to read bearer token for scraping targets.
+	BearerTokenFile string `json:"bearerTokenFile,omitempty"`
```
Same as TLSConfig

auhlig replied:
removed

brancz commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -542,6 +551,71 @@ type ServiceMonitorSpec struct {
+	HonorLabels bool `json:"honorLabels,omitempty"`
+	// BasicAuth allow an endpoint to authenticate over basic authentication
+	// More info: https://prometheus.io/docs/operating/configuration/#endpoints
+	BasicAuth *BasicAuth `json:"basicAuth,omitempty"`
```
Same as TLSConfig and BearerTokenFile

auhlig replied:
removed

auhlig commented:
Mh. Not sure why the CI runs into a timeout: https://travis-ci.org/coreos/prometheus-operator/jobs/526384616

brancz commented:
Interesting. First time I'm seeing this, and it doesn't look like master is having this problem. I think it might have to do with this PR. Have you tried running the latest state on a new cluster?

auhlig commented:
Strange: It times out for all commits but at different tests.
I tested the initial state in a cluster. Let me do that for the latest one as well.

metalmatze commented:
I don't think it matters at what tests it timeouts. Looking at the fact that it needs 50min and we have a limit of 55min for those tests, I suppose that something deadlocks and doesn't finish.
I would guess that this is a test in this PR, so worth trying those tests in isolation.

auhlig commented:
I'm convinced. Must be related. I look into it.
Thanks for the feedback so far

auhlig commented:
I cannot see what's causing the timeout in the e2e test. Would appreciate any hint @brancz @metalmatze @squat 🙏

paulfantom commented:
This needs to be regenerated. Please run `make generate-in-docker` and commit generated files from `example/` directory.

auhlig commented:
Thanks @paulfantom.
I fixed the e2e test and rebased. Noticed that I have to re-generate but was interrupted by a meeting. 
Will do in the evening, then this PR should be ready for a final review.

auhlig commented:
Done @paulfantom. All tests are green 🎉 
Could you take a look again?

paulfantom commented:
lgtm

brancz commented on Documentation/api.md:
```diff
@@ -267,6 +328,8 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | podMetadata | Standard object’s metadata. More info: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#metadata Metadata Labels and Annotations gets propagated to the prometheus pods. | *[metav1.ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta) | false |
 | serviceMonitorSelector | ServiceMonitors to be selected for target discovery. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
 | serviceMonitorNamespaceSelector | Namespaces to be selected for ServiceMonitor discovery. If nil, only check own namespace. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
+| podMonitorSelector | PodMonitors to be selected for target discovery. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
```
could we mark these experimental for now? I'd just like people to be able to use this for a while before we officially support this in a stable API

auhlig replied:
done

brancz commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,248 @@ func (cg *configGenerator) generateConfig(
+	// Relabel namespace and pod and service labels into proper labels.
+	relabelings = append(relabelings, []yaml.MapSlice{
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_node_name"}},
```
this is not something we currently do or encourage with ServiceMonitors, this should be done at query time against kube-state-metrics

auhlig replied:
makes sense. dropped this relabeling

brancz commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,248 @@ func (cg *configGenerator) generateConfig(
+			{Key: "target_label", Value: "pod"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_phase"}},
```
this would result in a new target when it shouldn't

brancz commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,248 @@ func (cg *configGenerator) generateConfig(
+			{Key: "target_label", Value: "pod_phase"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_ready"}},
```
this would result in a new target when it shouldn't

auhlig replied:
makes sense. dropped this relabeling

brancz commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,248 @@ func (cg *configGenerator) generateConfig(
+	// as a fallback the port number.
+
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_name"}},
```
I'm not sure this is a great default, job is intended to group workloads of the same type, this will cause distinct job labels for those. Maybe the generate name would be more appropriate? (not sure there is a meta label for those though)

auhlig replied:
Isn't this also true services if the LabelSelector of a ServiceMonitorSpec applies to multiple and differently names services? 

Which generated name did you have in mind? It doesn't look like theres a good pod meta label to solve this (1). 
---
(1) https://prometheus.io/docs/prometheus/latest/configuration/configuration/#pod

brancz replied:
Most pods are not created directly, and their controller sets the `.metadata.generateName` which is the "scheme" the pods name was chosen, often good enough for grouping, but probably a bit too magic for this. I'd say maybe let's just default to the PodMonitor's name as the job label?

auhlig replied:
Thanks for the explanation. 
Defaulting the job name to the PodMonitors key (`$namespace/$name`) instead.

auhlig commented:
rebased and fixed e2e test

auhlig commented:
If you're okay with https://github.com/coreos/prometheus-operator/pull/2566/commits/788a6b0c6536b6a3868995574bc5327d6f99842c, I'd rebase & resolve the conflicts again @brancz

brancz commented on pkg/prometheus/promcfg.go:
```diff
@@ -272,6 +294,235 @@ func (cg *configGenerator) generateConfig(
+			{Key: "target_label", Value: "namespace"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_container_name"}},
```
side note: we should add this in the ServiceMonitor as well :+1:

auhlig commented:
Rebased and resolved conflicts. LGTY @brancz?

brancz commented:
lgtm :+1: probably worth one more review as it's such a large one

@s-urbaniak @metalmatze @squat @paulfantom ?

auhlig commented:
Thanks @brancz @metalmatze @s-urbaniak @paulfantom @squat :tada
